File: test/ModularPipelines.UnitTests/Models/SkipDecisionTests.cs
using System.Runtime.CompilerServices;
using System.Text.Json;
using ModularPipelines.Models;

namespace ModularPipelines.UnitTests.Models;

public class SkipDecisionTests
{
    [Test]
    public async Task ImplicitConversions_AreNotExposed()
    {
        var conversions = typeof(SkipDecision)
            .GetMethods()
            .Where(method => method.Name == "op_Implicit")
            .Select(method =>
                $"{method.GetParameters().Single().ParameterType.FullName}->{method.ReturnType.FullName}")
            .ToArray();

        await Assert.That(conversions).IsEmpty();
    }

    [Test]
    public async Task ShouldSkip_IsInitOnly()
    {
        var setter = typeof(SkipDecision)
            .GetProperty(nameof(SkipDecision.ShouldSkip))!
            .SetMethod!;

        await Assert.That(setter.ReturnParameter.GetRequiredCustomModifiers())
            .Contains(typeof(IsExternalInit));
    }

    [Test]
    public async Task JsonRoundTrip_PreservesDecision()
    {
        var expected = SkipDecision.Skip("Serialized reason");

        var json = JsonSerializer.Serialize(expected);
        var actual = JsonSerializer.Deserialize<SkipDecision>(json);

        await Assert.That(actual).IsEqualTo(expected);
    }

    [Test]
    public async Task Skip()
    {
        var skipDecision = SkipDecision.Skip("Blah!");

        using (Assert.Multiple())
        {
            await Assert.That(skipDecision.ShouldSkip).IsTrue();
            await Assert.That(skipDecision.Reason).IsEqualTo("Blah!");
        }
    }

    [Test]
    public async Task DoNotSkip()
    {
        var skipDecision = SkipDecision.DoNotSkip;

        using (Assert.Multiple())
        {
            await Assert.That(skipDecision.ShouldSkip).IsFalse();
            await Assert.That(skipDecision.Reason).IsNull();
        }
    }

    [Test]
    [Arguments(true)]
    [Arguments(false)]
    public async Task When(bool shouldSkip)
    {
        var skipDecision = SkipDecision.When(shouldSkip, "Blah!");

        using (Assert.Multiple())
        {
            await Assert.That(skipDecision.ShouldSkip).IsEqualTo(shouldSkip);
            await Assert.That(skipDecision.Reason).IsEqualTo(shouldSkip ? "Blah!" : null);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Models/SkipDecisionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Models

open ModularPipelines.Models
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type SkipDecisionTests() =
    [<Test>]
    member _.True_Implicit_Cast() = async {
        let skipDecision: SkipDecision = true
        do! check(Assert.That(skipDecision.ShouldSkip).IsTrue())
        do! check(Assert.That(skipDecision.Reason).IsNull())
    }

    [<Test>]
    member _.String_Implicit_Cast() = async {
        let skipDecision: SkipDecision = "Foo!"
        do! check(Assert.That(skipDecision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(skipDecision.Reason), "Foo!"))
    }

    [<Test>]
    member _.False_Implicit_Cast() = async {
        let skipDecision: SkipDecision = false
        do! check(Assert.That(skipDecision.ShouldSkip).IsFalse())
        do! check(Assert.That(skipDecision.Reason).IsNull())
    }

    [<Test>]
    member _.Skip() = async {
        let skipDecision = SkipDecision.Skip("Blah!")
        do! check(Assert.That(skipDecision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(skipDecision.Reason), "Blah!"))
    }

    [<Test>]
    member _.DoNotSkip() = async {
        let skipDecision = SkipDecision.DoNotSkip
        do! check(Assert.That(skipDecision.ShouldSkip).IsFalse())
        do! check(Assert.That(skipDecision.Reason).IsNull())
    }

    [<Test>]
    [<Arguments(true)>]
    [<Arguments(false)>]
    member _.Of(shouldSkip: bool) = async {
        let skipDecision = SkipDecision.Of(shouldSkip, "Blah!")

        if shouldSkip then
            do! check(Assert.That(skipDecision.ShouldSkip).IsTrue())
            do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(skipDecision.Reason), "Blah!"))
        else
            do! check(Assert.That(skipDecision.ShouldSkip).IsFalse())
            do! check(Assert.That(skipDecision.Reason).IsNull())
    }
